### ISSUE TITLE
SqlException: Violation of PRIMARY KEY constraint

### ISSUE THREAD

**Issue by bleissem** [JonPSmith/EfCore.GenericServices#21]
Hello, 
first of all thanks for your work on EfCore.GenericServices, we are using it. 
Unfortunately there is a strange behavior when updating nested dto's 
I created a project for you at 
https://github.com/bleissem/TestNestedInDtosIssue13
and add a little description. So you should be able to reproduce the exception
SqlException: Violation of PRIMARY KEY constraint 'PK_AddressNotOwned'. Cannot insert duplicate key in object 'dbo.AddressNotOwned'. 

To reproduce the error it should be enough to deploy the database from the Database1 project. 
Then change the Connection string in the application.json and then run the WebApplication1 project.
In the ValuesController the code to examine is in the Get function 

Regards
Alex B.

**Comment by JonPSmith** [comment]
Hi @bleissem,

The error you show is a SQL exception and is about the primary key. I think this something about how you set the key. Have you tried writing the data out by hand, i.e. writing your own mapping code instead of using DTOs? 

By doing that you know if its a database setup problem or a DTO copy problem.

**Comment by bleissem** [comment]
Hello @JonPSmith ,

yes manual mapping will do. But that is not what I want to accomplish.
I just updated 
https://github.com/bleissem/TestNestedInDtosIssue13

There are now 2 Controller Actions:
http://localhost:5000/api/values/worksgreat
http://localhost:5000/api/values/doesnotwork
As the name says one works, the other does not. 

Feel free to look at the code 
https://github.com/bleissem/TestNestedInDtosIssue13/blob/master/WebApplication1/Controllers/ValuesController.cs

Under the hood the funcion 
```
            public TEntity ReturnExistingEntity(object[] keys)
            {
                return _context.Set<TEntity>().Find(keys);
            }
```
does not load navigation properties when calling doesnotwork function, wheras in the worksgreat function the function ReturnExistingEntity does load navigation properties.

**Comment by tesnep** [comment]
@bleissem If you change using DTO to EF class 
`InContactAddressDto updateDTO = service.ReadSingle<InContactAddressDto>(w => w.ContactAddressId == contactAddressId)` 

To

`var updateEF= service.ReadSingle<InContactAddress>(w => w.ContactAddressId == contactAddressId)`
and create a new InContactAddressDto updateDTO to map from updateEF, does it work?

**Comment by bleissem** [comment]
Hi @tesnep ,
thanks for your suggestion. 
Not using DTO's is something I don't want to accomplish. 
In my real world szenario there a certain mappings that are done with the PerDtoConfig class.
Also the DTO object is passed into the controller action by the frontend.
So not using DTO's is not an option for me. 

The example I provided is just related to [a unit test](https://github.com/bleissem/EfCore.GenericServices/blob/master/Tests/UnitTests/GenericServicesPublic/TestNestedInDtosIssue13.cs) that is capabable of saving nested DTO.

**Comment by JonPSmith** [comment]
Hi @bleissem,

I have added a unit test part to your application (pushed to master so you can get it) and I can't see anything wrong with GenericServices. I wonder if your dapper code isn't working properly??

Anyway, its not GenericServices so I am closing this issue.

**Comment by JonPSmith** [comment]
Sorry - push to the GitHub wasn't allowed. Here is my unit test

```csharp
sing System;
using System.Linq;
using ClassLibrary1;
using ClassLibrary1.DTOs;
using GenericServices.PublicButHidden;
using GenericServices.Setup;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using TestSupport.EfHelpers;
using WebApplication1.Controllers;
using Xunit;
using Xunit.Extensions.AssertExtensions;

namespace Test
{
    public class UnitTest1
    {

        [Fact]
        public void TestCreateContactAddressOk()
        {
            //SETUP
            var options = SqliteInMemory.CreateOptions<MyDBContext>();
            using (var context = new MyDBContext(options))
            {
                context.Database.EnsureCreated();


                //ATTEMPT
                context.Add(new ContactAddress
                    {Name = "unit test", AddressNotOwned = new AddressNotOwned {Address1 = "some street"}});
                context.SaveChanges();
            }
            using (var context = new MyDBContext(options))
            {
                //VERIFY
                var cAddr = context.ContactAddress.Include(x => x.AddressNotOwned).Single();
                cAddr.AddressNotOwned.ShouldNotBeNull();
            }
        }

        [Fact]
        public void TestMapContactAddressOk()
        {
            //SETUP
            var options = SqliteInMemory.CreateOptions<MyDBContext>();
            using (var context = new MyDBContext(options))
            {
                context.Database.EnsureCreated();
                context.Add(new ContactAddress
                    { Name = "unit test", AddressNotOwned = new AddressNotOwned { Address1 = "some street" } });
                context.SaveChanges();
            }
            using (var context = new MyDBContext(options))
            {
                var utData = context.SetupSingleDtoAndEntities<InContactAddressDto>();
                var service = new CrudServices(context, utData.ConfigAndMapper);

                //ATTEMPT
                var result = service.ReadSingle<InContactAddressDto>(1);

                //VERIFY
                result.Addess.ShouldNotBeNull();
            }
        }


        [Fact]
        public void TestCallControllerMethodDoesNotWorkOk()
        {
            //SETUP
            var options = SqliteInMemory.CreateOptions<MyDBContext>();
            using (var context = new MyDBContext(options))
            {
                context.Database.EnsureCreated();
                context.Add(new ContactAddress
                    { Name = "unit test", AddressNotOwned = new AddressNotOwned { Address1 = "some street" } });
                context.SaveChanges();
            }
            using (var context = new MyDBContext(options))
            {
                var utData = context.SetupSingleDtoAndEntities<InContactAddressDto>();
                var service = new CrudServices(context, utData.ConfigAndMapper);
                var controller = new ValuesController();

                //ATTEMPT
                var result = controller.DoesNotWorkWithoutDapper(service, context, 1);

                //VERIFY
                result.Value.Addess.ShouldNotBeNull();
            }
        }
    }
}
```

**Comment by bleissem** [comment]
Hi @JonPSmith ,
thanks for providing code to the testproject. I added you as a Collaborator.
Fortunately your comit 
Commit e71c02fe
made it to the Repo.

To show the error on calling the UpdateAndSave funtion, I just did another commit 
Commit 6ce7148e
to the 
TestCallControllerMethodDoesNotWorkOk 
function.
When calling 
service.UpdateAndSave(toBeUpdated);

there is an exception being thrown 
AutoMapper.AutoMapperMappingException: 'Error mapping types.'

Well saving a new DTO works but updating an exsisting DTO doesn't seem to be working.

**Comment by JonPSmith** [comment]
OK. Two things.

1. The error is a AutoMapper error so you need to look at how to solve that. 
2. If you solved the AutoMapper problem I would still recommend using EF Core directly because you are saving a relationship where one is new (ContactAddress), and the relationship (AddressNotOwned) was read in (i.e. tracked) as part of GenericServices's ReadSingle. That's pretty complex, and did you mean that? For anything complex I use EF Core directly.

**Comment by bleissem** [comment]
Hi @JonPSmith ,
i've been off a few days and didn't response.
I just updated 
https://github.com/bleissem/TestNestedInDtosIssue13
and adjusted some things to avoid the automapper errormessage.
I think the automapper related errors are now gone.
I also include the GenericServices Code to be able to show you exactly when the test fails and when it does not fail.

For the test:
TestCallControllerMethodDoesNotWorkOk
now the test still fails but this time with:
Message: Microsoft.EntityFrameworkCore.DbUpdateException : An error occurred while updating the entries. See the inner exception for details.
---- Microsoft.Data.Sqlite.SqliteException : SQLite Error 19: 'UNIQUE constraint failed: AddressNotOwned.Id'.
This is because when updating an entity there are no navigation properties beeing loaded.

In commit [46c088560e313f5c63fd0ed56d90fd69888eabe5](https://github.com/bleissem/TestNestedInDtosIssue13/commit/46c088560e313f5c63fd0ed56d90fd69888eabe5)
I modified the function CreateMapper.ReturnExistingEntity in GenericServices so that updating actually works and the unit test passes.
However the unit test: TestCallControllerMethodDoesNotWorkOk fails when using the original code from GenericServices.
as seen in commit [95287da18ed115215df73896b90a71668073090e](https://github.com/bleissem/TestNestedInDtosIssue13/commit/95287da18ed115215df73896b90a71668073090e)
